### PR TITLE
docs warnings as errors and initial page similar to README

### LINKED ISSUE #432: Testing docs and doc tests
Originally reported by: **Piotr Banaszkiewicz (BitBucket: [pbanaszkiewicz](http://bitbucket.org/pbanaszkiewicz), GitHub: [pbanaszkiewicz](http://github.com/pbanaszkiewicz))**

----------------------------------------

Testing docs with tox right now requires `make`. Not everyone has it installed.
(And `regen` target for tox doesn't work, too)

I propose to expand testing docs and doctests so that:

1. rst syntax in docstrings is checked
2. `make` is no longed needed
3. doc trees are shared per tox env. run
4. `rm -rf` is not needed ([lets not repeat other projects' mistakes](https://github.com/MrMEEE/bumblebee-Old-and-abbandoned/commit/a047be85247755cdbe0acce6f1dafc8beb84f2ac#diff-3fbb47e318cd8802bd325e7da9aaabe8L351))
5. Sphinx is more strict about warnings (`-W`)

I'm ready to submit a pull request if you agree.

I would also like to somehow force developers to write docs about implemented new features or changed pytest behavior.

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/432

### LINKED ISSUE #1112: Reorganize content page
I always found the pytest docs quite hard to navigate - and because of someone with the same issue in IRC just now I realized why: The stuff which is probably most relevant (`assert.rst` and `fixture.rst`) is hidden on `examples.rst` but not mentioned at all in `content.rst`.

Some other stuff in `content.rst` is probably not that relevant... I think the whole thing should have some reorganization.

I'll take a look in a few weeks (still have a growing todo list), but if someone else wants to pick this up, please go ahead!

### LINKED ISSUE #1708: New main page now that docs reside in doc.pytest.org
Now that http://doc.pytest.org is officially up, I propose we change our `pytest.org` landing page. 

pytest.org currently is the same `index` page for the docs, but I see that other projects usually follow a different direction, containing a short project description, a short snippet right at the beginning and proper links to the rest of the documentation, contact, etc etc. I would say pretty much like our current `README`.

Examples:

* http://www.fabfile.org (docs: http://docs.fabfile.org)
* http://flask.pocoo.org (docs: http://jinja.pocoo.org/docs)
* http://jinja.pocoo.org (docs: http://flask.pocoo.org/docs)

In order to do that, we would probably move a few things in `docs`, but I propose to keep `index.rst` as the main *docs* page, and create another `landing-page.rst` or something like this (suggestions welcome), and only it would be uploaded to `pytest.org`, with proper links to http://doc.pytest.org.

I think it would be really nice to have this for 3.0.

cc @pfctdayelise @hackebrot

### LINKED ISSUE #1829: Issues when building docs
`thm` reported some issues in `#pylib` when building the docs:

- [ ] Missing targets in the 2.9.1 and 2.9.2 annoucements:

```
.../doc/en/announce/release-2.9.1.rst:39: ERROR: Unknown target name: "@nicoddemus".
.../doc/en/announce/release-2.9.1.rst:42: ERROR: Unknown target name: "@nicoddemus".
.../doc/en/announce/release-2.9.1.rst:46: ERROR: Unknown target name: "@tomviner".
.../doc/en/announce/release-2.9.1.rst:49: ERROR: Unknown target name: "@nicoddemus".
.../doc/en/announce/release-2.9.2.rst:42: ERROR: Unknown target name: "@ronnypfannschmidt".
.../doc/en/announce/release-2.9.2.rst:45: ERROR: Unknown target name: "@tomviner".
.../doc/en/announce/release-2.9.2.rst:56: ERROR: Unknown target name: "@pquentin".
.../doc/en/announce/release-2.9.2.rst:56: ERROR: Unknown target name: "@ronnypfannschmidt".
.../doc/en/announce/release-2.9.2.rst:60: ERROR: Unknown target name: "@omarkohl".
.../doc/en/announce/release-2.9.2.rst:63: ERROR: Unknown target name: "@hackebrot".
```

- [ ] `.../doc/en/announce/sprint2016.rst:42: ERROR: Unexpected indentation.`
- [ ] Imports which were moved in 3.0 but not adjusted:

```
.../doc/en/builtin.rst:61: WARNING: autodoc: failed to import function u'fixture' from module u'_pytest.python'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 525, in import_object
    obj = self.get_attr(obj, part)
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 421, in get_attr
    return safe_getattr(obj, name, *defargs)
  File "/usr/lib/python2.7/site-packages/sphinx/util/inspect.py", line 115, in safe_getattr
    raise AttributeError(name)
AttributeError: fixture
.../doc/en/builtin.rst:67: WARNING: autodoc: failed to import class u'FixtureRequest' from module u'_pytest.python'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 525, in import_object
    obj = self.get_attr(obj, part)
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 421, in get_attr
    return safe_getattr(obj, name, *defargs)
  File "/usr/lib/python2.7/site-packages/sphinx/util/inspect.py", line 115, in safe_getattr
    raise AttributeError(name)
AttributeError: FixtureRequest

.../doc/en/plugins.rst:134: WARNING: failed to import _pytest.pdb

.../doc/en/writing_plugins.rst:626: WARNING: autodoc: failed to import class u'FixtureDef' from module u'_pytest.python'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 525, in import_object
    obj = self.get_attr(obj, part)
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 421, in get_attr
    return safe_getattr(obj, name, *defargs)
  File "/usr/lib/python2.7/site-packages/sphinx/util/inspect.py", line 115, in safe_getattr
    raise AttributeError(name)
AttributeError: FixtureDef
```

- [ ] `.../doc/en/contents.rst:10: WARNING: toctree contains reference to nonexisting document u'apiref'`
- [ ] `.../doc/en/doctest.rst:19: WARNING: Literal block ends without a blank line; unexpected unindent.`
- [ ] `.../doc/en/feedback.rst:8: WARNING: malformed hyperlink target.`
- [ ] `.../doc/en/feedback.rst:4: ERROR: Unknown target name: "py.test is pretty much the best thing ever".`
- [ ] `.../doc/en/monkeypatch.rst:81: WARNING: missing attribute replace in object _pytest.monkeypatch.MonkeyPatch`
- [ ] `.../_pytest/hookspec.py:docstring of _pytest.hookspec.pytest_assertrepr_compare:0: WARNING: duplicate object description of _pytest.hookspec.pytest_assertrepr_compare, other instance in .../doc/en/assert.rst, use :noindex: for one of them`
- [ ] `ImportError` for `pluggy`:

```
.../doc/en/writing_plugins.rst:647: WARNING: autodoc: failed to import class u'PluginManager' from module u'pluggy'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 518, in import_object
    __import__(self.modname)
ImportError: No module named pluggy
```

- [ ] documents not referred from anywhere?

```
checking consistency... .../doc/en/adopt.rst:: WARNING: document isn't included in any toctree
.../doc/en/announce/release-3.0.0.rst:: WARNING: document isn't included in any toctree
.../doc/en/assert.rst:: WARNING: document isn't included in any toctree
.../doc/en/bash-completion.rst:: WARNING: document isn't included in any toctree
.../doc/en/builtin.rst:: WARNING: document isn't included in any toctree
.../doc/en/contact.rst:: WARNING: document isn't included in any toctree
.../doc/en/customize.rst:: WARNING: document isn't included in any toctree
.../doc/en/doctest.rst:: WARNING: document isn't included in any toctree
.../doc/en/feedback.rst:: WARNING: document isn't included in any toctree
.../doc/en/fixture.rst:: WARNING: document isn't included in any toctree
.../doc/en/index.rst:: WARNING: document isn't included in any toctree
.../doc/en/license.rst:: WARNING: document isn't included in any toctree
.../doc/en/mark.rst:: WARNING: document isn't included in any toctree
.../doc/en/parametrize.rst:: WARNING: document isn't included in any toctree
.../doc/en/proposals/parametrize_with_fixtures.rst:: WARNING: document isn't included in any toctree
.../doc/en/skipping.rst:: WARNING: document isn't included in any toctree
.../doc/en/status.rst:: WARNING: document isn't included in any toctree
.../doc/en/unittest.rst:: WARNING: document isn't included in any toctree
.../doc/en/writing_plugins.rst:: WARNING: document isn't included in any toctree
.../doc/en/xdist.rst:: WARNING: document isn't included in any toctree
.../doc/en/xunit_setup.rst:: WARNING: document isn't included in any toctree
.../doc/en/yieldfixture.rst:: WARNING: document isn't included in any toctree
```

- [ ] Undefined labels:

```
.../doc/en/getting-started.rst:32: WARNING: undefined label: installation issues (if the link has no caption the label must precede a section header)
.../doc/en/getting-started.rst:199: WARNING: undefined label: apiref (if the link has no caption the label must precede a section header)
```

- [ ] `copying static files... WARNING: html_static_path entry u'.../doc/en/_static' does not exist`

- [ ] failing highlighting:

```
writing output... [ 68%] example/reportingdemo
.../doc/en/example/reportingdemo.rst:12: WARNING: Could not lex literal_block as "python". Highlighting skipped.
writing output... [ 69%] example/simple
.../doc/en/example/simple.rst:40: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:62: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:109: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:157: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:171: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:207: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:285: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:309: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:322: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:355: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:417: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:488: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:592: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:621: WARNING: Could not lex literal_block as "python". Highlighting skipped.
.../doc/en/example/simple.rst:682: WARNING: Could not lex literal_block as "python". Highlighting skipped.
```

### PR DESCRIPTION
**Pull request by nicoddemus**
* Changed "index" to be more similar to the README (#1708).
* Fixes numerous issues like missing documents, syntax errors, etc (#1829, #432).
* Mention all docs in "contents.rst" so it's easier for users to locate (#1112).
* Add doc generation and checking to Travis and AppVeyor, to avoid re-introducing errors.

I pushed this branch to pytest's repository instead of mine so readthedocs would build the documentation, which can be [browsed here](http://doc.pytest.org/en/doc-warnings/).

Fixes #432, Fixes #1112, Fixes #1708, Fixes #1829

### CODE REVIEW

File: _pytest/python.py [hunk 1]
```diff
@@ -711,31 +711,26 @@ class Metafunc(fixtures.FuncargnamesCompatAttr):
     They help to inspect a test function and to generate tests according to
     test configuration or values specified in the class or module where a
     test function is defined.
-
-    :ivar fixturenames: set of fixture names required by the test function
```
Comment (nicoddemus) [inline _pytest/python.py]: I had to change from docstring to inline comments because of sphinx-doc/sphinx#2549. :disappointed:

File: doc/en/contents.rst [hunk 2]
```diff
@@ -10,32 +10,51 @@ Full pytest documentation
 .. toctree::
    :maxdepth: 2
 
-   overview
-   apiref
-   example/index
+   getting-started
```
Comment (nicoddemus) [inline doc/en/contents.rst]: Suggestions on how to order these topics are welcome.

**Comment by nicoddemus** [comment]
After fixing the numerous errors in the documentation (missing links, syntax errors, etc) it's clear to me that treating documentation warnings as errors is absolutely vital to keep the documentation healthy. :sweat_smile:

File: doc/en/announce/sprint2016.rst [hunk 3]
```diff
@@ -39,8 +39,6 @@ Over 20 participants took part from 4 continents, including employees
 from Splunk, Personalkollen, Cobe.io, FanDuel and Dolby. Some newcomers
 mixed with developers who have worked on pytest since its beginning, and
 of course everyone in between.
-    Ana Ribeiro, Brazil
```
Comment (RonnyPfannschmidt) [inline doc/en/announce/sprint2016.rst]: where did we go ^^

File: doc/en/contents.rst (hunk 2 above)
Comment (RonnyPfannschmidt) [inline doc/en/contents.rst]: i recall @pfctdayelise working on the structure, perhaps we can use the pictures and merge the docs branch 

also how far are those appart

File: doc/en/announce/sprint2016.rst (hunk 3 above)
Comment (nicoddemus) [inline doc/en/announce/sprint2016.rst]: I think it was decided to remove the entire participants list at some point, but those two names seem to be have left by accident (they were a syntax error in Sphinx)

File: doc/en/index.rst [hunk 4]
```diff
@@ -1,59 +1,91 @@
+:orphan:
 
 .. _features:
 
 pytest: helps you write better programs
-=============================================
```
Comment (RonnyPfannschmidt) [inline doc/en/index.rst]: why this one ?

File: doc/en/contents.rst (hunk 2 above)
Comment (nicoddemus) [inline doc/en/contents.rst]: Not sure, I thought they meant a complete overhaul of the docs, including rewriting sections. Here I was interested in getting zero warnings, and to do that I had to include all documents in the table of contents, but which is something we already had agreed would be for the best (flat better than nested, etc)

File: doc/en/overview.rst [hunk 5]
```diff
@@ -1,13 +0,0 @@
-==================================================
```
Comment (RonnyPfannschmidt) [inline doc/en/overview.rst]: something is strange here - why are those gone as well

File: doc/en/index.rst (hunk 4 above)
Comment (nicoddemus) [inline doc/en/index.rst]: Please see #1708.

File: doc/en/overview.rst (hunk 5 above)
Comment (nicoddemus) [inline doc/en/overview.rst]: Those were all moved to the main "contents.rst"... having this in a nested TOC didn't seem to bring much to the table. All those documents can  be found directly in the contents page now.

Comment (RonnyPfannschmidt) [inline doc/en/overview.rst]: i see, i understand now

**Comment by nicoddemus** [comment]
Perhaps @hackebrot and @pfctdayelise would like to take a look as well? :grin:

File: doc/en/getting-started.rst [hunk 6]
```diff
@@ -19,17 +19,15 @@ Installation and Getting Started
...
-    pip install -U pytest # or
-    easy_install -U pytest
+    pip install -U pytest
 
 To check your installation has installed the correct version::
 
     $ pytest --version
     This is pytest version 3.0.0, imported from $PYTHON_PREFIX/lib/python3.5/site-packages/pytest.py
 
-If you get an error checkout :ref:`installation issues`.
```
Comment (The-Compiler) [inline doc/en/getting-started.rst]: Does this refer to an existing document? If so, that reference seems useful to me?

File: doc/en/index.rst [hunk 7]
```diff
@@ -1,59 +1,91 @@
...
+        def test_answer():
+    >       assert func(3) == 5
+    E       assert 4 == 5
+    E        +  where 4 = func(3)
+
+    test_sample.py:5: AssertionError
+    ======= 1 failed in 0.12 seconds ========
+
+
+Due to ``py.test``'s detailed assertion introspection, only plain ``assert`` statements are used.
```
Comment (The-Compiler) [inline doc/en/index.rst]: Shouldn't this be `pytest`? #dropthedot

File: doc/en/xdist.rst [hunk 8]
```diff
@@ -1,197 +0,0 @@
-
-.. _`xdist`:
```
Comment (The-Compiler) [inline doc/en/xdist.rst]: I'm guessing this was all removed because it's sufficiently documented in pytest-xdist's docs?

File: tox.ini [hunk 9]
```diff
@@ -93,15 +93,15 @@ commands=
...
-deps=sphinx
-     PyYAML
+deps=
+    sphinx
+    PyYAML
 
 commands=
-    make clean
-    make html
+    sphinx-build -W -b html . _build
```
Comment (The-Compiler) [inline tox.ini]: Should we still keep `doc/en/Makefile`?

File: doc/en/getting-started.rst (hunk 6 above)
Comment (nicoddemus) [inline doc/en/getting-started.rst]: It doesn't... I didn't really track where it might have went, but I remember there being something related to weird Windows installation problems because "python" wasn't on the PATH... don't think it is really relevant if that's what it was.

File: doc/en/index.rst (hunk 7 above)
Comment (nicoddemus) [inline doc/en/index.rst]: Heh good point

File: doc/en/xdist.rst (hunk 8 above)
Comment (nicoddemus) [inline doc/en/xdist.rst]: Exactly, it was a (very old) copy of xdist's README here.

File: tox.ini (hunk 9 above)
Comment (nicoddemus) [inline tox.ini]: Good point, I think we should drop it and suggest using `tox -e docs` instead.

File: doc/en/index.rst (hunk 7 above)
Comment (nicoddemus) [inline doc/en/index.rst]: Done :grin:

File: tox.ini (hunk 9 above)
Comment (nicoddemus) [inline tox.ini]: Hmm just realized we still depend on it because of `regendoc`. 

Probably it would be better to get rid of it and move pertinent tasks to `tox`, but we can do that in a separate PR.

**Comment by nicoddemus** [comment]
Pushed last comments made by @The-Compiler.

**Comment by hackebrot** [comment]
Hey! 👋 

I'd love to review this, but I don't have the time today. 😞

File: .travis.yml [hunk 10]
```diff
@@ -26,6 +26,7 @@ env:
     - TESTENV=py27-nobyte
     - TESTENV=doctesting
     - TESTENV=freeze
+    - TESTENV=docs
```
Comment (hackebrot) [inline .travis.yml]: 👍

File: doc/en/adopt.rst [hunk 11]
```diff
@@ -1,3 +1,4 @@
+:orphan:
```
Comment (hackebrot) [inline doc/en/adopt.rst]: Might be good to add a comment:

From the sphinx docs:
```
If set, warnings about this file not being included in any toctree will be suppressed.
```

File: doc/en/assert.rst [hunk 12]
```diff
@@ -208,6 +208,7 @@ It is possible to add your own detailed explanations by implementing
 the ``pytest_assertrepr_compare`` hook.
 
 .. autofunction:: _pytest.hookspec.pytest_assertrepr_compare
+   :noindex:
```
Comment (hackebrot) [inline doc/en/assert.rst]: I suspect this caused a warning?

File: doc/en/conf.py [hunk 13]
```diff
@@ -144,7 +144,7 @@
 # Add any paths that contain custom static files (such as style sheets) here,
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
-html_static_path = ['_static']
+# html_static_path = ['_static']
```
Comment (hackebrot) [inline doc/en/conf.py]: We don't need this anymore?

File: doc/en/example/simple.rst [hunk 14]
```diff
@@ -296,7 +314,9 @@ which will add the string to the test header accordingly::
 You can also return a list of strings which will be considered as several
 lines of information.  You can of course also make the amount of reporting
 information on e.g. the value of ``config.getoption('verbose')`` so that
-you present more information appropriately::
+you present more information appropriately:
```
Comment (hackebrot) [inline doc/en/example/simple.rst]: I find this sentence slightly confusing, might be out-of-scope for this PR, but I think we should reword it.

File: doc/en/assert.rst (hunk 12 above)
Comment (nicoddemus) [inline doc/en/assert.rst]: Yes, because it had the same directive elsewhere and Sphinx complained about which one to add to its main index.

File: doc/en/getting-started.rst [hunk 15]
```diff
@@ -19,17 +19,15 @@ Installation and Getting Started
 Installation
 ----------------------------------------
 
-Installation options::
+Installation::
 
-    pip install -U pytest # or
-    easy_install -U pytest
+    pip install -U pytest
```
Comment (hackebrot) [inline doc/en/getting-started.rst]: 👍

File: doc/en/conf.py (hunk 13 above)
Comment (nicoddemus) [inline doc/en/conf.py]: There's no `_static` directory so it was generating a warnings.

File: doc/en/index.rst [hunk 16]
```diff
@@ -1,59 +1,91 @@
...
+
+    def test_answer():
+        assert func(3) == 5
+
+
+To execute it::
+
+    $ pytest
+    ======= test session starts ========
+    platform linux -- Python 3.4.3, pytest-2.8.5, py-1.4.31, pluggy-0.3.1
```
Comment (hackebrot) [inline doc/en/index.rst]: This should be Python 3.5.0 and more importantly pytest-3.0.1 😁

File: doc/en/adopt.rst (hunk 11 above)
Comment (nicoddemus) [inline doc/en/adopt.rst]: No problem

File: doc/en/index.rst [hunk 17]
```diff
@@ -1,59 +1,91 @@
...
+
+
+Features
+--------
+
+- Detailed info on failing :ref:`assert statements <assert>` (no need to remember ``self.assert*`` names);
+
+- :ref:`Auto-discovery <test discovery>` of test modules and functions;
+
+- :ref:`Modular fixtures <fixture>`  for
```
Comment (hackebrot) [inline doc/en/index.rst]: I think there is an extra space right before ``for``

File: doc/en/example/simple.rst (hunk 14 above)
Comment (nicoddemus) [inline doc/en/example/simple.rst]: Could you suggest something you find more adequate? I would be happy to change it

File: tox.ini (hunk 9 above)
Comment (hackebrot) [inline tox.ini]: Sounds sensible

File: doc/en/index.rst (hunk 16 above)
Comment (nicoddemus) [inline doc/en/index.rst]: True! I will execute regendocs on it.

Unfortunately our README contains the same text and is not updated automatically. I wonder if it's important to display that banner at all on the README and if just removing it would suffice for demonstration purposes.

File: doc/en/assert.rst (hunk 12 above)
Comment (hackebrot) [inline doc/en/assert.rst]: ok, fair enough

File: doc/en/conf.py (hunk 13 above)
Comment (hackebrot) [inline doc/en/conf.py]: ok, cool

File: doc/en/example/simple.rst (hunk 14 above)
Comment (hackebrot) [inline doc/en/example/simple.rst]: I would, if only I could understand what it's supposed to mean. 😕

File: doc/en/index.rst (hunk 16 above)
Comment (hackebrot) [inline doc/en/index.rst]: I always remove it from blog posts. It's not really helping at this point.

W/e works best in this case!

Comment (nicoddemus) [inline doc/en/index.rst]: You mean this:

```
$ pytest
======= test session starts ========
collected 1 items

test_sample.py F

======= FAILURES ========
_______ test_answer ________

    def test_answer():
>       assert func(3) == 5
E       assert 4 == 5
E        +  where 4 = func(3)

test_sample.py:5: AssertionError
======= 1 failed in 0.12 seconds ========
```

File: doc/en/index.rst (hunk 17 above)
Comment (nicoddemus) [inline doc/en/index.rst]: :+1:

File: doc/en/index.rst (hunk 16 above)
Comment (hackebrot) [inline doc/en/index.rst]: Yup.

Comment (nicoddemus) [inline doc/en/index.rst]: Will do, also for the README

**Comment by hackebrot** [comment]
Great work @nicoddemus! 👏
